### PR TITLE
SOL-150799 (T5 prep): extract internal/idpclient

### PR DESCRIPTION
**Pull request by amitmorade**
## Summary

Refactor pre-work for the upcoming SOL-150799 T5 RFC 8693 token-exchange Exchanger. Extracts the existing IdP HTTP client construction (currently inside `internal/auth/middleware.go`) into a new `internal/idpclient` package so the future Hop 2 consumer can share the same TLS root configuration and SOL-150219 timeout bound that Hop 1 already uses.

Pre-work only — no T5 functionality lands here. The Exchanger itself ships in a follow-up PR.

Jira: [SOL-150799](https://sol-jira.atlassian.net/browse/SOL-150799)

## Type of Change

- [x] Refactoring (no functional changes)

## Motivation and Context

T5 (the RFC 8693 token-exchange Exchanger) needs to talk to the same IdP that the OIDC verifier in `internal/auth` already talks to. The IdP HTTP client carries two non-trivial properties that the team has invested in:

1. **TLS root configuration with `SSL_CERT_FILE` escape hatch** — needed for corporate-CA deployments on macOS/Windows where the OS-native trust store ignores `SSL_CERT_FILE`.
2. **The SOL-150219 timeout bound** — `http.Client.Timeout` is the only thing that survives go-oidc's `WithoutCancel` strip and reaches lazy JWKS refresh.

If T5 builds its own client, both properties drift the moment one side changes. The fix is to extract the construction into a shared package and have both hops call the same recipe.

Doing this extraction as its own PR ahead of T5 keeps the T5 review focused on Exchanger semantics rather than HTTP-client plumbing.

### Commit 1: `extract internal/idpclient package` ([255618c](https://github.com/SolaceDev/solace-broker-mcp/pull/CURRENT/commits/255618c))

- New `internal/idpclient` package with `NewHTTPClient(opts ...Option)`.
  - Unexported `newBaseTransport` carries the shared TLS root setup; `NewHTTPClient` wraps it with the configured timeout. This names the seam where a future `NewMTLSHTTPClient` sibling will plug in — mTLS authenticates only the IdP's token endpoint, and JWKS is always public, so Hop 1 will never need a client cert.
  - Unexported `httpClientTimeout` default (`defaults.DefaultOIDCHTTPTimeout`); no exported writable global.
  - Exported `WithTimeout` functional option — production callers pass no options; tests pass `WithTimeout(500ms)` for SOL-150219 regression coverage in bounded wall-clock.
- Removed `oidcHTTPClient` and `oidcHTTPClientTimeout` from `internal/auth/middleware.go`.
- `NewAuthMiddleware` and `NewTokenVerifier` grow a single `*http.Client` parameter, **nil-default**:
  - Production callers pass `nil` and `createOIDCTokenVerifier` builds the client lazily via `idpclient.NewHTTPClient()` — preserves the pre-PR "build at point of use, only when needed" property.
  - Disabled and static modes never touch IdP code paths; `cmd/server/main.go` does not construct any IdP client.
  - Tests pass a non-nil client built via `idpclient.NewHTTPClient(idpclient.WithTimeout(...))` to control the timeout.
- Table-driven SSL_CERT_FILE branch test moved to `internal/idpclient/client_test.go`, joined by a `WithTimeout`-behavior test.
- The OIDC `NewTokenVerifier` TLS-integration sub-tests stay in `internal/auth/middleware_test.go`, renamed to `Test_NewTokenVerifier_TLSWithSSLCertFile` to reflect their narrowed scope.

### Commit 2: `move JWKS-refresh timeout tests to test/integration/` ([8258ed0](https://github.com/SolaceDev/solace-broker-mcp/pull/CURRENT/commits/8258ed0))

- `Test_OIDCJWKSRefreshTimeout` and `Test_OIDCJWKSRefreshTimeout_SSLCertFile` move from `internal/auth/middleware_test.go` to a new `test/integration/oidc_jwks_refresh_timeout_test.go`.
- These exercise the auth middleware + OIDC token verifier + go-oidc together against a hung fake IdP — multi-component coverage that per `test/integration/README.md` belongs in this tier.
- Pure relocation. Tests already used `idpclient.NewHTTPClient(idpclient.WithTimeout(500ms))` in commit 1, so the move is mechanical via the exported `auth.NewAuthMiddleware` / `idpclient.NewHTTPClient` surface.
- Took the move as an opportunity to lean on the standard library where the original code reached past it: `server.Certificate()` in place of dereferencing `server.TLS.Certificates[0].Certificate[0]` and calling `x509.ParseCertificate`, and `map[string]string` in place of `map[string]interface{}` for the OIDC discovery JSON.

## Testing

**Test Environment:**
- Go version: 1.23
- OS: macOS (darwin/arm64); CI covers Linux
- Broker version: n/a (no broker interaction in this PR)

**Tests Performed:**
- [x] Unit tests added/updated — `internal/idpclient/client_test.go` covers the five SSL_CERT_FILE branches plus `WithTimeout` behavior; the integration sub-tests in `internal/auth/middleware_test.go` were narrowed and renamed.
- [x] Integration tests added/updated — the two JWKS-refresh timeout tests now live in `test/integration/`.
- [x] Tested locally with `go test -race ./...` — all packages green.
- [ ] Tested with real Solace broker — not relevant (no broker code paths touched).

**Test Coverage:**
- SSL_CERT_FILE branches: env unset / nonexistent file / unreadable file / non-PEM contents / valid PEM, plus SOL-150219 default-bound invariant.
- `WithTimeout`: default applies when no option, override wins when option passed.
- SOL-150219 end-to-end on both the default-trust-store path and the SSL_CERT_FILE corporate-CA path (the latter was the branch the original fix left unbounded).
- OIDC `NewTokenVerifier` TLS handshake with and without the corporate CA trusted.

## Breaking Changes

None for external consumers of the binary.

The signature change on `NewAuthMiddleware` and `NewTokenVerifier` is internal to this repo. All in-repo callers (one in `cmd/server/main.go`, one in `test/integration/raw_subject_token_capture_test.go`, plus the relocated and existing tests in `internal/auth/middleware_test.go`) are updated in this PR.

### Code Quality
- [x] Code follows project coding standards
- [x] Self-review completed
- [x] Code is well-commented (explains "why", not "what")
- [x] No commented-out code or debug statements left in

### Security
- [x] No credentials in code
- [x] `/check-logs` clean on the full PR diff — zero new log calls introduced in production code (`idpclient` is silent by design)
- [x] No security vulnerabilities introduced

### Testing
- [x] All tests pass locally: `go test -race ./...`
- [x] No race conditions detected
- [x] Edge cases covered by tests
- [x] Error paths tested
- [x] Test names clearly describe what is being tested

### Documentation
- [ ] README.md updated — not applicable
- [ ] `docs/` updated — not applicable (no architecture change)
- [x] godoc comments added/updated for exported functions

### Git & CI
- [x] Commits have clear, descriptive messages
- [x] No merge conflicts (branch built on top of latest `main`)
- [ ] CI checks passing — pending CI run on this branch

## Design Journey (for context)

Two design questions surfaced during this PR worth recording so reviewers don't have to retrace them:

**1. Why functional options on `NewHTTPClient` instead of an exported `HTTPClientTimeout` var?**

Initial draft exported the timeout var so external tests could mutate it. Code review correctly flagged this as a design regression — an exported writable global is mutable by any package in the build, no synchronization, no compile-time protection. Functional options give the test the override capability through a typed, side-effect-isolated API. Production callers pass no options.

**2. Why nil-default `*http.Client` on `NewAuthMiddleware`/`NewTokenVerifier` instead of always injecting at `main`?**

An intermediate iteration had `main` build the client unconditionally and pass it in. That introduced a behavior regression: `disabled` and `static` modes — which never talk to an IdP — would now read `SSL_CERT_FILE` at startup and could fail on a malformed env var they had no reason to touch. The nil-default shape preserves the pre-PR "build at point of use, only when needed" property while still giving tests the injection seam they need.

## Follow-Up

T5 (the Exchanger itself) will land in a separate PR. It imports `idpclient` and calls `idpclient.NewHTTPClient()` exactly the way `createOIDCTokenVerifier` does today. When mTLS support arrives, a sibling `idpclient.NewMTLSHTTPClient(cert)` plugs in via `newBaseTransport` without disturbing existing callers.

## Related Issues/PRs

- Related to T5 of the SOL-150070 OAuth Token Exchange decomposition (see `docs/oauth/token-exchange-SOL-150070/decomposition.md`).
- Builds on [#111 (SOL-150797 T3)](https://github.com/SolaceDev/solace-broker-mcp/pull/111).

---

**For Reviewers:**

**Focus Areas:**
1. The nil-default contract on `NewAuthMiddleware` / `NewTokenVerifier` — is the doc comment clear enough?
2. The SOL-150219 timeout invariant is preserved end-to-end — verify by reading `internal/idpclient/client.go` (constructor sets `Timeout`) and `test/integration/oidc_jwks_refresh_timeout_test.go` (end-to-end coverage on both default and SSL_CERT_FILE paths).
3. The split between unit tests in `internal/idpclient/` (per-branch SSL_CERT_FILE behavior) and integration tests in `test/integration/` (multi-component propagation) matches the tier convention in `test/integration/README.md`.

**Questions:** None outstanding. 

[SOL-150799]: https://sol-jira.atlassian.net/browse/SOL-150799?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: internal/idpclient/client.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+// newBaseTransport returns a fresh *http.Transport carrying the shared TLS
+// root configuration: the system trust store plus, when SSL_CERT_FILE is
+// set, the PEM bundle that file points at. On macOS and Windows, Go's
+// default TLS verification delegates to the OS-native trust store and
+// ignores SSL_CERT_FILE; an explicit RootCAs pool bypasses that
+// delegation so the env var works consistently on every platform.
+func newBaseTransport() (*http.Transport, error) {
+	t := http.DefaultTransport.(*http.Transport).Clone()
+
```
[the remark this answers is not in the document]
Comment (amitmorade) [reply internal/idpclient/client.go:75]: Good catch — but this code is pre-existing. The oidcHTTPClient function on main has the same http.DefaultTransport.(*http.Transport).Clone() line at internal/auth/middleware.go:219. We moved it verbatim. Filing as a follow-up rather than expanding this PR's scope.

File: internal/auth/middleware.go [hunk 2]
```diff
@@ -113,19 +113,20 @@ func createStaticTokenVerifier(expectedToken string) sdkauth.TokenVerifier {
...
+// discovery call.
+func createOIDCTokenVerifier(cfg *config.ServerConfig, httpClient *http.Client) (sdkauth.TokenVerifier, error) {
+	if httpClient == nil {
+		c, err := idpclient.NewHTTPClient()
+		if err != nil {
+			return nil, err
+		}
+		httpClient = c
 	}
 	clientCtx := oidc.ClientContext(context.Background(), httpClient)
```
[the remark this answers is not in the document]
Comment (amitmorade) [reply internal/auth/middleware.go:131]: Real failure mode — a hand-rolled client with `Timeout == 0` passed here would bring back the hung refresh.

I think the fix fits better in `idpclient` than here, though. The companion comment on `NewHTTPClient` rejects bad timeouts at the source, so every client built through the package is already bounded. The only way a bad one reaches this function is if someone skips `idpclient` and builds their own — and a `Timeout > 0` check here wouldn't reliably catch that anyway (`24h` passes the check but is basically unbounded).

We did consider the stronger option: have `idpclient` return its own opaque type (a wrapper around `*http.Client` that only its constructor can produce) and make this function accept that instead of a raw `*http.Client`. That would make an unvalidated client impossible to pass — the compiler would enforce it. We decided against it because it's a fair amount of plumbing (wrapper type, an accessor for the go-oidc seam that needs the raw client, threading it through every caller) to guard an invariant whose worst case is a degraded timeout rather than a security issue — and it's the kind of rigor that's better spent on the TLS-verification path than on the timeout. The nil-default + producer-side validation already gives us a bounded client on every real path.

Open to revisiting if we think the stronger guarantee is worth that cost

File: internal/idpclient/client.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+	cfg := &config{timeout: defaults.DefaultOIDCHTTPTimeout}
+	for _, opt := range opts {
+		opt(cfg)
+	}
+	transport, err := newBaseTransport()
+	if err != nil {
+		return nil, err
+	}
+	return &http.Client{Transport: transport, Timeout: cfg.timeout}, nil
+}
```
[the remark this answers is not in the document]
Comment (amitmorade) [reply internal/idpclient/client.go:119]: Addressed in 92568da: NewHTTPClient now returns `idpclient: timeout must be positive, got <d>` when the resolved timeout is `<= 0`. Validation lives in the constructor rather than in WithTimeout itself to match the convention used by major Go libraries (grpc-go, prometheus/client_golang) — Option functions stay as simple setters; the constructor applies policy. Two test cases cover the zero and negative paths.

File: test/integration/oidc_jwks_refresh_timeout_test.go [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+	// JWKS client validate TLS. Without this the verifier could not be built.
+	caPath := filepath.Join(t.TempDir(), "ca.crt")
+	if err := os.WriteFile(
+		caPath,
+		pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: server.Certificate().Raw}),
+		0600,
+	); err != nil {
+		t.Fatalf("failed to write CA file: %v", err)
+	}
+	t.Setenv("SSL_CERT_FILE", caPath)
```
Comment (solace-aross) [inline test/integration/oidc_jwks_refresh_timeout_test.go:209]: The test-isolation guard that protects this test didn't travel with it into `test/integration/`.

The auth package's `TestMain` (`internal/auth/middleware_test.go:189`) deliberately fires the `crypto/x509` system-roots `sync.Once` with `SSL_CERT_FILE` cleared, so a test pointing `SSL_CERT_FILE` at a test CA can't leak that CA into the process-wide root cache and break sibling assertions. Its comment notes the leak only surfaces on Linux CI.

Here, `t.Setenv("SSL_CERT_FILE", caPath)` followed by `idpclient.NewHTTPClient` → `x509.SystemCertPool()` seeds that cache with the test CA. `t.Setenv` restores the env var after the test, but it does not reset the `sync.Once` — the cached roots persist for the rest of the test binary. `test/integration/` has no `TestMain` to prime or clear it.

Nothing breaks today, because no other integration test asserts an untrusted TLS server is rejected — so this is a latent footgun rather than a live bug. But the next integration test that relies on system roots (or on `SSL_CERT_FILE` being unset) inherits an order-dependent Linux-CI flake, the exact failure the auth package already paid to fix. Worth adding the same `TestMain` guard to this package now, or at least a comment here flagging the hazard so the next author isn't surprised.

File: internal/idpclient/client.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (solace-aross) [reply internal/idpclient/client.go:75]: Agree it's pre-existing and a fair scope call. One nuance pushes the other way, though: the assertion now lives in the package whose docstring explicitly invites more consumers (T5, a future `NewMTLSHTTPClient`), and `NewHTTPClient`/`newBaseTransport` already return `error`. So the safe-assertion fix is ~3 lines onto an error path that already exists, in the one place that's now the shared source of truth — a different cost/benefit than fixing it inline in a single private function.

Not insisting. If it stays a follow-up, dropping the ticket reference in this thread would keep it from getting lost.

File: test/integration/oidc_jwks_refresh_timeout_test.go [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+// (SOL-150219). Both the default-trust-store path and the SSL_CERT_FILE
+// corporate-CA path are covered, because the original fix left the
+// custom-CA branch unbounded.
+//
+// These compose internal/auth's middleware, the OIDC token verifier, and
+// the go-oidc library against a fake hung IdP — the assertion only makes
+// sense with all three layers wired together, so per the test/integration/
+// README they belong in this tier rather than next to a single component's
+// unit tests.
+package integration
```
Comment (solace-awaheed) [inline test/integration/oidc_jwks_refresh_timeout_test.go]: Should this be package integration_test as well? To keep naming consistent with other files

File: test/integration/oidc_jwks_refresh_timeout_test.go (hunk 4 above)
Comment (solace-awaheed) [reply test/integration/oidc_jwks_refresh_timeout_test.go:209]: Flagged on my end as well

Comment (amitmorade) [reply test/integration/oidc_jwks_refresh_timeout_test.go:209]: Good catch — addressed in `e67d54b`. Added `test/integration/main_test.go` with a `TestMain` that fires `x509.SystemCertPool()` with `SSL_CERT_FILE` / `SSL_CERT_DIR` cleared before any test in the package runs, mirroring the guard in `internal/auth/middleware_test.go`. The two `TestMain`s stay in sync because they protect the same hazard from the same shared dependency on `x509.SystemCertPool`. The leak is now structurally prevented for any future OAuth-flow test added to this tier, not just documented.

File: internal/idpclient/client.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (amitmorade) [reply internal/idpclient/client.go:75]: Persuaded — addressed in a984f7e. The cost/benefit argument lands: the assertion now lives in the shared source of truth that explicitly invites more consumers (T5, future NewMTLSHTTPClient), and newBaseTransport already returns error, so the fix is ~3 lines on an existing error path.

newBaseTransport now uses the comma-ok form and returns "idpclient: http.DefaultTransport is not *http.Transport (got <type>); cannot construct IdP HTTP client" instead of panicking with runtime.TypeAssertionError.

Test: a fakeRoundTripper satisfies http.RoundTripper without being *http.Transport. The test swaps and restores http.DefaultTransport under a package-level mutex (defaultTransportMu), with restore-then-unlock fused in a single defer so the lock isn't released while the fake is in place. Asserts NewHTTPClient returns the error and a nil client.

File: test/integration/oidc_jwks_refresh_timeout_test.go (hunk 5 above)
Comment (amitmorade) [reply test/integration/oidc_jwks_refresh_timeout_test.go]: Agreed — addressed in `be0f741`. Normalized all four files in `test/integration/` to `package integration_test`. The inconsistency originated in `8258ed0` when I created `oidc_jwks_refresh_timeout_test.go` with the wrong package name; `e67d54b` (the `TestMain` add) then inherited the same mistake.

Verified the rename is safe: no test in the directory referenced unexported symbols across the in-package/external-package boundary, nothing outside `test/integration/` imports the package, and both `package integration` and `package integration_test` compile into the same test binary so `TestMain` still gates the suite. Pure cosmetic rename.

**Comment by solace-lootii** [comment]
Mostly LGTM. but One thing I'm not sure about and wanted to ask:

`internal/idpclient` looks like the only package touching `x509.SystemCertPool` that *didn't* get the `TestMain` cert-pool guard that `internal/auth` and `test/integration/` now have. I'm wondering if that matters, because:

- `client_test.go`'s `t.Setenv("SSL_CERT_FILE", …)` + `NewHTTPClient` reaches `x509.SystemCertPool()` ([client.go:84](internal/idpclient/client.go#L84)), which fires the process-wide `sync.Once`.
- `t.Setenv` restores the env var but **not** the `Once` so whatever `SSL_CERT_FILE` was in effect when it first fired stays cached for the rest of the test binary.

I think it's **not** an actual problem today if the value that fires it is invalid PEM, so it can't grant trust, and nothing here asserts TLS rejection. But **is this a valid concern later**, once a `NewMTLSHTTPClient` (or other TLS-sensitive) test lands in this same package? Or am I wrong here and there's a reason `idpclient` doesn't need the guard the other two got? 

Also +1 on the injected-`Timeout == 0` thread,  producer-side check is the right depth.

**Comment by amitmorade** [comment]
> Mostly LGTM. but One thing I'm not sure about and wanted to ask:
> 
> `internal/idpclient` looks like the only package touching `x509.SystemCertPool` that _didn't_ get the `TestMain` cert-pool guard that `internal/auth` and `test/integration/` now have. I'm wondering if that matters, because:
> 
> * `client_test.go`'s `t.Setenv("SSL_CERT_FILE", …)` + `NewHTTPClient` reaches `x509.SystemCertPool()` ([client.go:84](internal/idpclient/client.go#L84)), which fires the process-wide `sync.Once`.
> * `t.Setenv` restores the env var but **not** the `Once` so whatever `SSL_CERT_FILE` was in effect when it first fired stays cached for the rest of the test binary.
> 
> I think it's **not** an actual problem today if the value that fires it is invalid PEM, so it can't grant trust, and nothing here asserts TLS rejection. But **is this a valid concern later**, once a `NewMTLSHTTPClient` (or other TLS-sensitive) test lands in this same package? Or am I wrong here and there's a reason `idpclient` doesn't need the guard the other two got?
> 
> Also +1 on the injected-`Timeout == 0` thread, producer-side check is the right depth.

Good catch — addressed in `2177027`. Added `internal/idpclient/main_test.go` with the same `TestMain` pattern that `internal/auth/` and `test/integration/` already have. You were right that `client_test.go`'s `TestNewHTTPClient_SSLCertFile` "valid PEM" case is the third place that points `SSL_CERT_FILE` at a test CA and triggers `SystemCertPool()`, and we hadn't primed this package's `sync.Once`.

Today nothing breaks because no other test in the package asserts an untrusted-server rejection, but the next TLS-sensitive test added here (e.g. the future `NewMTLSHTTPClient` the package docstring invites) would have inherited the same order-dependent Linux-CI flake we already paid to fix twice elsewhere. The three `TestMain`s now stay in sync because they protect the same hazard from the same shared dependency.

Also added a pointer comment above `TestNewHTTPClient_SSLCertFile` directing future readers to `main_test.go` so the "why" is discoverable from the test that needs the guard.

And confirming on the injected-`Timeout == 0` thread — agreed, producer-side check is the right depth, and we explicitly considered (and deferred) the stronger typed-wrapper option for the same reason.

**Review by solace-lootii — APPROVED** [review]
nice, everything else LGTM!

**Review by solace-awaheed — APPROVED** [review]
LGTM

**Review by solace-aross — APPROVED** [review]
Approving. Clean extraction, and the iteration on this thread was a pleasure to read. Every Copilot and reviewer point landed somewhere defensible: timeout validation at the constructor (the right depth), the comma-ok transport assertion now living in the shared source of truth, and three `TestMain` guards kept in sync against the same `x509.SystemCertPool` hazard.

Both design-journey calls are right and well recorded for the T5 follow-up: functional options over an exported timeout var, and the nil-default `*http.Client` that preserves build-at-point-of-use for `disabled`/`static` modes.

Verified locally: build, vet, and the `idpclient` + `auth` unit tests pass. No functional change, two peer approvals already. Nothing blocking from me.

Minor, optional: the body still reads "Marking draft," but the PR isn't a draft and has approvals. Worth tidying that line before merge so the history reads straight.
